source/game.production.script.bmx

```diff
@@ -126,30 +126,51 @@ Type TScriptCollection Extends TGameObjectCollection
 
 	Method GenerateRandom:TScript(avoidTemplateIDs:Int[])
 		Local template:TScriptTemplate
-		If Not avoidTemplateIDs Or avoidTemplateIDs.length = 0
-			template = GetScriptTemplateCollection().GetRandomByFilter(True, True)
-		Else
-			Local foundValid:Int = False
-			Local tries:Int = 0
-
-			template = GetScriptTemplateCollection().GetRandomByFilter(True, True, "", avoidTemplateIDs)
-			'get a random one, ignore avoid IDs
-			If Not template And avoidTemplateIDs And avoidTemplateIDs.length > 0
-				Print "TScriptCollection.GenerateRandom() - warning. No available template found (avoid-list too big?). Trying an avoided entry."
+		'iterate several times in case template with production limit and protected title is involved
+		For Local i:Int = 0 Until 20
+			If Not avoidTemplateIDs Or avoidTemplateIDs.length = 0
 				template = GetScriptTemplateCollection().GetRandomByFilter(True, True)
+			Else
+				Local foundValid:Int = False
+				Local tries:Int = 0
+	
+				template = GetScriptTemplateCollection().GetRandomByFilter(True, True, "", avoidTemplateIDs)
+				'get a random one, ignore avoid IDs
+				If Not template And avoidTemplateIDs And avoidTemplateIDs.length > 0
+					Print "TScriptCollection.GenerateRandom() - warning. No available template found (avoid-list too big?). Trying an avoided entry."
+					template = GetScriptTemplateCollection().GetRandomByFilter(True, True)
+				EndIf
+				'get a random one, ignore availability
+				If Not template
+					Print "TScriptCollection.GenerateRandom() - failed. No available template found (avoid-list too big?). Using an unfiltered entry."
+					template = GetScriptTemplateCollection().GetRandomByFilter(False, True)
+				EndIf
 			EndIf
-			'get a random one, ignore availability
-			If Not template
-				Print "TScriptCollection.GenerateRandom() - failed. No available template found (avoid-list too big?). Using an unfiltered entry."
-				template = GetScriptTemplateCollection().GetRandomByFilter(False, True)
-			EndIf
-		EndIf
 
-		Local script:TScript = TScript.CreateFromTemplate(template,True)
-		script.SetOwner(TOwnedGameObject.OWNER_NOBODY)
-		Add(script)
+			'If the template has a production limit, ensure no protected title is used
+			If template.GetProductionLimitMax() > 1
+				Local newTitleOk:Int = False
```

Comment:
Theoretisch kann auch auf die verzichtet werden, wenn dafuer "tempName" im Fall von "nicht OK" nullgesetzt wuerde - und man am Ende macht:
`If not tempName Then template = Null`

Alternativ wuerde man ueberlegen koennen, ob "GenerateFinalTitle()" umgeschrieben werden soll um "title:TLocalizedString" als Argument/Parameter zu akzeptieren - und wenn das nicht Null ist, wird eben das befuellt.
Also das nur, falls man nicht "bis zu 20" temporaere TLocalizedStrings anlegen will. Dann muss aber im "GenerateFinalTitle()" die uebergebene Variable zurueckgesetzt werden ("geleert").

Reply:
An GenerateFinalTitle will ich eigentlich nichts ändern. Das Ersetzen der "Merkvariable" kann ich machen. Das macht es tatsächlich leichter lesbar.